server/storage/rethinkdb.go

```diff
@@ -294,7 +329,110 @@ func (rdb RethinkDB) CheckHealth() error {
 	return nil
 }
 
-// GetChanges is not implemented for RethinkDB
+func (rdb RethinkDB) writeChange(gun string, version int, sha256, category string) error {
+	now := time.Now()
+	ch := Change{
+		CreatedAt: now,
+		GUN:       gun,
+		Version:   version,
+		SHA256:    sha256,
+		Category:  category,
+	}
+	_, err := gorethink.DB(rdb.dbName).Table(ch.TableName()).Insert(
+		ch,
+		gorethink.InsertOpts{
+			Conflict: "error", // default but explicit for clarity of intent
+		},
+	).RunWrite(rdb.sess)
+	return err
+}
+
+// GetChanges returns up to pageSize changes starting from changeID. It uses the
+// blackout to account for RethinkDB's eventual consistency model
 func (rdb RethinkDB) GetChanges(changeID string, pageSize int, filterName string) ([]Change, error) {
-	return nil, errors.New("Not Implemented")
+	var (
+		lower, upper, bound []interface{}
+		idx                 = "rdb_created_at_id"
+		max                 = []interface{}{gorethink.Now().Sub(blackoutTime), gorethink.MaxVal}
+		min                 = []interface{}{gorethink.MinVal, gorethink.MinVal}
+		order               gorethink.OrderByOpts
+		reversed            bool
+	)
+	if filterName != "" {
+		idx = "rdb_gun_created_at_id"
+		max = append([]interface{}{filterName}, max...)
+		min = append([]interface{}{filterName}, min...)
+	}
+
+	switch changeID {
+	case "0", "-1":
+		lower = min
+		upper = max
+	default:
+		bound, idx = rdb.bound(changeID, filterName)
+		if pageSize < 0 {
+			lower = min
+			upper = bound
+		} else {
+			lower = bound
+			upper = max
+		}
+	}
+
+	if changeID == "-1" || pageSize < 0 {
+		reversed = true
+		order = gorethink.OrderByOpts{Index: gorethink.Desc(idx)}
+	} else {
+		order = gorethink.OrderByOpts{Index: gorethink.Asc(idx)}
+	}
+
+	if pageSize < 0 {
+		pageSize = pageSize * -1
+	}
+
+	changes := make([]Change, 0, pageSize)
+
+	// Between returns a slice of results from the rethinkdb table.
+	// The results are ordered using BetweenOpts.Index, which will
+	// default to the index of the immediately preceding OrderBy.
+	// The lower and upper are the start and end points for the slice
+	// and the Left/RightBound values determine whether the lower and
+	// upper values are included in the result per normal set semantics
+	// of "open" and "closed"
```

Comment:
"open" means including the values, right?

Reply:
"open" means excluding values.